### PR TITLE
CP-18062: Ensure one PVS cache VDI is created per-site and per-host

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

**Comment by robhoes** [comment]
This looks like it's on the right track, but will need some tweaks. Also, two of the unit tests are failing. Since sharad is away on Monday, I'll have a look at fixing it myself.

File: ocaml/xapi/xapi_pvs_cache.ml [hunk 1]
```diff
@@ -21,19 +21,22 @@ exception No_cache_sr_available
...
+	let find_all ~__context ~sr ~site =
 		let open Db_filter_types in
-		Db.VDI.get_refs_where ~__context ~expr:(And
+		match Db.PVS_cache_storage.get_refs_where ~__context ~expr:(And
 			(Eq (Field "SR", Literal (Ref.string_of sr)),
-			Eq (Field "type", Literal "pvs_cache")))
+			Eq (Field "site", Literal (Ref.string_of site))))
+		with
+		| [] -> None, []
+		| cacheSR :: _ -> Some cacheSR, (Db.PVS_cache_storage.get_cache_VDIs ~__context ~self:cacheSR)
```
Comment (lindig) [inline ocaml/xapi/xapi_pvs_cache.ml:31]: I'm a little bit surprised that the VDIs for the other matches (in `_`) are not returned. Maybe a comment would make help to understand this better, especially when the function is called `find_all`.

File: ocaml/xapi/xapi_pvs_cache.ml [hunk 2]
```diff
@@ -21,19 +21,22 @@ exception No_cache_sr_available
...
 
-	let find_one ~__context ~sr =
-		match find_all ~__context ~sr with
-		| [] -> None
-		| vdi :: _ -> Some vdi
+	let find_one ~__context ~sr ~site ~host =
+		match find_all ~__context ~sr ~site with
+		| (None, _) -> None
+		| (_, vdi_map) -> try Some (List.assoc host vdi_map) with _ -> None
```
Comment (lindig) [inline ocaml/xapi/xapi_pvs_cache.ml:36]: I think this should match only `Not_found` to avoid that we are masking errors here.

**Comment by robhoes** [comment]
I have updated an merged this outside this PR.

**Pull request closed without merging**